### ISSUE TITLE
socket.error: [Errno 21] Invalid argument (Bash on windows, Linux kernels < 3.9)

### ISSUE THREAD

**Issue by Ciberth** [cs01/gdbgui#36]
Hi there, 

I wanted to try gdbgui with bash (ubuntu) on windows. I installed gdb, pip and afterwards the gdbgui. I downloaded the hello.c code and used make to compile it. 

When trying to run gdb hello I see the output down below. Any ideas what this might be and how I can solve it?

![gdbgui](https://cloud.githubusercontent.com/assets/1833112/23586453/5e83f03a-0195-11e7-8822-e99518889de4.png)

**Comment by cs01** [comment]
It looks like it's related to the `eventlet` async library being used for websockets. You might want to try switching the socketio mode from `eventlet` to `gevent`. In the file [gdbgui/backend.py](https://github.com/cs01/gdbgui/blob/fe225403770e3fff9b6cfb21a72ece45fd51a232/gdbgui/backend.py#L47)

change this line

    socketio = SocketIO(async_mode='eventlet')

to 

    socketio = SocketIO(async_mode='gevent')

That's just a shot in the dark, but let me know if it helps.

What version of ubuntu are you using and how do you have it set up in windows? You might want to try using docker. Someone on reddit said they were successfully using gdbgui with docker.

**Comment by Ciberth** [comment]
I wanted to explicit try it on the ubuntu version from the anniversary build of windows ;) 

lsb_release -a gives:

> No LSB modules are available.
> Distributor ID: Ubuntu
> Description:    Ubuntu 14.04.5 LTS
> Release:        14.04
> Codename:       trusty

For myself (and you) I took the liberty of documenting my steps.

I tried installing gevent with pip and easy_install:

> apt-get install libevent-dev
> apt-get install python-all-dev
> easy_install greenlet
> easy_install gevent 

but got a lot of warnings (and some errors) honestly, it even hanged while installing in both cases. 

> sudo apt-get install python-gevent

did the trick, I think (?)

Then I edited the file like you said:

> vim /usr/local/lib/python2.7/dist-packages/gdbgui/backend.py

I ran gdbgui hello.c and it **resolved** the problem. Nicely done!

I'm gonna test it now a bit more, hope this was useful. Keep up the good work!

**Comment by ssokolow** [comment]
> and how do you have it set up in windows?
> I wanted to explicit try it on the ubuntu version from the anniversary build of windows

Sounds like @Ciberth is using the [Windows Subsystem for Linux](https://en.wikipedia.org/wiki/Windows_Subsystem_for_Linux) (WSL). Here's Microsoft's About page for it:

https://msdn.microsoft.com/en-us/commandline/wsl/about

**TL;DR:** The Windows NT kernel isn't actually tied to the Win32 API. It was explicitly developed with the intent of supporting the Win32, OS/2, and POSIX APIs as equal siblings via "subsystems" (that's also why NTFS has a POSIX mode which allows characters like `:` in filenames).

Over the course of the NT kernel's lifetime, Microsoft paired the POSIX-side support with various userlands (eg. Interix) and called it [Windows Services for UNIX](https://en.wikipedia.org/wiki/Windows_Services_for_UNIX). That got deprecated in Windows 8 and removed in Windows 8.1 and WSL is the effective successor.

**Comment by cs01** [comment]
@Ciberth I believe some recent updates have made the workaround unnecessary. Can you confirm, and if so close this ticket?

**Comment by Ciberth** [comment]
Sorry for the late reply. 
I don't have the luxury to test it cleanly from the ground up (I have some conflicts in python installs which I can not resolve untill I break something else).

 **If someone else has a clean windows setup to test it, please go ahead and let us know so I can close the issue.**

**Comment by cs01** [comment]
I am going to close out since I have confirmed it works on the v3.2 of the Linux kernel
